# <span style="color:green"> Start Here </span> 

[Please Complete the following form with your details]

Student Name - Alexandra Pinto<br>
Student id - 20211599<br>
Contact e-mail - 20211599@novaims.unl.pt<br>


Student Name - Marco Galão<br>
Student id - 20201545<br>
Contact e-mail - r20201545@novaims.unl.pt<br>

# <span style="color:brown"> Part 1 -  Variable Declaration and Manipulation </span> 

## Exercise I - of Lists and Random numbers

**1**: Declare a variable X that stores a list of 100 integers randomly sampled between -100 and 100.<br>
Note: You are not allowed to use third party libraries -- such as Numpy, Pandas, or Scipy -- in this exercise.

In [1]:
#Needed import to generate random integers
import random 

#Creating a list X 
X = [random.randint(-100,100) for i in range(100)]

In [2]:
#Checking the created list
print(X)

[15, -16, -7, 17, -55, 72, -63, 45, 83, -91, 86, 42, -62, 99, 54, -96, -90, 49, -96, 88, -34, 18, -75, 16, 46, -44, -67, 41, 17, 31, -68, 100, -82, 72, -21, 86, 91, -18, 17, -100, -37, -36, -100, -65, -6, -48, -60, 27, 97, -8, -6, -95, 38, -87, -57, 97, 99, -88, -95, -57, 74, 59, -4, -95, 90, 40, -100, 2, 24, 37, -94, 19, -86, 34, -25, 88, -27, -22, 47, -46, -47, 46, -95, 13, -45, 43, 57, -52, -67, 34, 79, 69, -37, -96, 54, 92, -91, 9, 11, 97]


**2**: How many odd numbers are in list X?


In [3]:
# Initializing a counter for the odd numbers
odd_count = 0

#Creating a for loop to access to every number in the list and check if its a odd number or no
for i in X:
    #Creating an if statement, when the number leaves a remainder when divided by 2 it means that is an odd number
    if i % 2 != 0:
        # Incrementing the ood counter by 1 each time an odd number is encountered
        odd_count += 1 

#Printing to achieve the answer to the question
print("Number of odd numbers in list X:", odd_count)

Number of odd numbers in list X: 50


**2.1** Check if there are as many even numbers as there are odd numbers, else discard the list and generate a new one.

In [4]:
# We will adapt the code in the previous exercise 
# Initializing a counter for the even numbers
even_count = 0

#Creating a for loop to access to every number in the list and check if its a even number or no
for i in X:
    #Creating an if statement, when the number does not leaves a remainder when divided by 2 it means that is an even number
    if i % 2 == 0:
        # Incrementing the ood counter by 1 each time an odd number is encountered
        even_count += 1 

In [5]:
#Now we will compare both counters

if odd_count == even_count:
    print("The list X has equal number of even and odd numbers")
else:
    print("The list X has different number of even and odd numbers, so we will generate a new list X")
    #Using the code created in the first exercise
    #Initializing an empty list X
    X = []
    #Creating a for loop to generate the 100 random integers
    for i in range(100):
        #In the following line we will generate a random integer in the asked interval
        random_int = random.randint(-100,100)
        #Now that the number was created we will append to the list created X
        X.append(random_int)

The list X has equal number of even and odd numbers


**2.2** Create a pipeline to automatically perform the tasks above such that you avoid having to run the cells multiple times to reach the desired outcome.

In [6]:
#Small note: in the following code we will use docstrings
"""
Docstrings give to the executer a brief description of the code that 
explains what it does, its inputs(args) and outputs(returns)
"""
import random
def generate_list():
    """
    Generates a list of 100 random integers between -100 and 100.
    
    Returns:
        list: A list of 100 random integers
    """
    return [random.randint(-100,100) for i in range(100)]

def count_even_odd_numbers(X):
    """
    Counts the number of even and odd numbers in the list
    
    Args:
        X (list): A list of integers
    
    Returns:
        tuple: A tuple contaning the count of even numbers and the count of odd numbers
    """
    even_count = 0
    odd_count = 0
    for i in X:
        #Creating an if statement, when the number does not leaves a remainder when divided by 2 it means that is an even number
        if i % 2 == 0:
            # Incrementing the ood counter by 1 each time an odd number is encountered
            even_count += 1  
        else:
            odd_count += 1
        
    return  even_count, odd_count

def pipeline():
    """
    Automates the task to generate a list.
    """
    #Using the first function created to generate the X list
    X= generate_list()
    #Calculate the number of even and odd numbers
    even_count, odd_count= count_even_odd_numbers(X)
    while even_count != odd_count:
        print("The list X has different number of even and odd numbers, so we will generate a new list X.")
        X= generate_list()
        even_count, odd_count= count_even_odd_numbers(X)
    print("The list X has equal number of even and odd numbers")
    print("List X:", X)
    
pipeline()


The list X has different number of even and odd numbers, so we will generate a new list X.
The list X has different number of even and odd numbers, so we will generate a new list X.
The list X has different number of even and odd numbers, so we will generate a new list X.
The list X has equal number of even and odd numbers
List X: [99, 96, -43, 19, -26, -7, -67, 5, 56, 52, 15, 97, -34, 58, -56, -92, -13, 41, -16, 10, -40, -88, -37, 52, 5, 36, -11, 49, 60, -80, 87, 64, 32, 88, 78, 51, -70, 57, 28, 77, 79, 46, -18, 26, 44, -31, 89, -31, -88, 38, 41, -13, -82, -48, 57, 33, -47, 31, -8, -43, 32, -20, 90, 57, 98, 48, -14, -36, -83, -12, 1, 19, 17, 38, 89, 37, 38, -24, -82, 38, -74, -55, 96, 9, -25, 0, -77, 39, -39, -47, -63, -77, 12, -27, -81, 12, 49, -20, 55, 89]


**3**: Print the number of digits that the 5th and 100th element of the list have. <br>
*Note: For instance, the number 1 contains one digit, the number 10 contains two digits, the number -2 contains one digit.*

In [7]:
fifth_element= X[4] #since the first number starts in 0 index
hundredth_element=X[99]

#To access the number of digits we wil use a combination of three functions:
#abs: returns the absolute value of the number
#str: converts the number into a string for counting the len
#len: returns the length of the string
print(f"The 5th element is {fifth_element} and has {len(str(abs(fifth_element)))} digits.")
print(f"The 100th element is {hundredth_element} and has {len(str(abs(hundredth_element)))} digits.")

The 5th element is -55 and has 2 digits.
The 100th element is 97 and has 2 digits.


**4**: Is the sum total of all the numbers in the list even or odd?

In [8]:
#Storing the sum of all numbers in the list
total_sum= sum(X)

#Check if the sum is even or odd
if total_sum % 2 == 0:
    print(f"The sum of all the numbers in list X is even and it is equal to {total_sum}.")
else:
    print(f"The sum of all the numbers in list X is odd and it is equal to {total_sum}.")

The sum of all the numbers in list X is even and it is equal to -298.


**5**: What is the average of the list X?

In [9]:
#Calculate the average of the list
avg = sum(X) / len(X)

print(f"The average of the list X is equal to {avg}")

The average of the list X is equal to -2.98


**5.1.** What is the population standard deviation? <br>

In [10]:
## WITHOUT USING MATH IMPORT
#Calculating the squared diff from the avg
squared_diffs = [(x - avg)**2 for x in X]

#Calculating the variance
var = sum(squared_diffs) /len(X)

#Calculate the pop. standard deviation
std_dev = var ** 0.5
##OR WITH MATH
#std_dev = math.sqrt(var)

print(f"The population standard deviation of X list is {std_dev}.")


The population standard deviation of X list is 63.916661364623856.


**6**: Sort list X in descending order and store the result in variable Xsort. <br>

In [11]:
#Using sorted function to sort X, and setting reverse=True to have it in descending order
Xsort = sorted(X, reverse=True)

#Check that actually work
print(X)
print(Xsort)

[15, -16, -7, 17, -55, 72, -63, 45, 83, -91, 86, 42, -62, 99, 54, -96, -90, 49, -96, 88, -34, 18, -75, 16, 46, -44, -67, 41, 17, 31, -68, 100, -82, 72, -21, 86, 91, -18, 17, -100, -37, -36, -100, -65, -6, -48, -60, 27, 97, -8, -6, -95, 38, -87, -57, 97, 99, -88, -95, -57, 74, 59, -4, -95, 90, 40, -100, 2, 24, 37, -94, 19, -86, 34, -25, 88, -27, -22, 47, -46, -47, 46, -95, 13, -45, 43, 57, -52, -67, 34, 79, 69, -37, -96, 54, 92, -91, 9, 11, 97]
[100, 99, 99, 97, 97, 97, 92, 91, 90, 88, 88, 86, 86, 83, 79, 74, 72, 72, 69, 59, 57, 54, 54, 49, 47, 46, 46, 45, 43, 42, 41, 40, 38, 37, 34, 34, 31, 27, 24, 19, 18, 17, 17, 17, 16, 15, 13, 11, 9, 2, -4, -6, -6, -7, -8, -16, -18, -21, -22, -25, -27, -34, -36, -37, -37, -44, -45, -46, -47, -48, -52, -55, -57, -57, -60, -62, -63, -65, -67, -67, -68, -75, -82, -86, -87, -88, -90, -91, -91, -94, -95, -95, -95, -95, -96, -96, -96, -100, -100, -100]


**6.1** Then replace each value in Xsort with index i as the sum of the values with index i-1 and i. <br> 

In [12]:
Xsort= [Xsort[0]] + [ Xsort[i-1]+ Xsort[i] for i in range(1, len(Xsort))]
Xsort

[100,
 199,
 198,
 196,
 194,
 194,
 189,
 183,
 181,
 178,
 176,
 174,
 172,
 169,
 162,
 153,
 146,
 144,
 141,
 128,
 116,
 111,
 108,
 103,
 96,
 93,
 92,
 91,
 88,
 85,
 83,
 81,
 78,
 75,
 71,
 68,
 65,
 58,
 51,
 43,
 37,
 35,
 34,
 34,
 33,
 31,
 28,
 24,
 20,
 11,
 -2,
 -10,
 -12,
 -13,
 -15,
 -24,
 -34,
 -39,
 -43,
 -47,
 -52,
 -61,
 -70,
 -73,
 -74,
 -81,
 -89,
 -91,
 -93,
 -95,
 -100,
 -107,
 -112,
 -114,
 -117,
 -122,
 -125,
 -128,
 -132,
 -134,
 -135,
 -143,
 -157,
 -168,
 -173,
 -175,
 -178,
 -181,
 -182,
 -185,
 -189,
 -190,
 -190,
 -190,
 -191,
 -192,
 -192,
 -196,
 -200,
 -200]

## Exercise II - The Most Uncool Band

**7**: Consider the dictionaries  *songs* and *users* that are declared in the cell below. <br>
Create a list with the names of the people who listened to more than one song. Print the List. <br>

In [13]:
songs = {
    "4589": "Mr. Blue Sky, Calling America, Twilight, Night in the City",
    "7312": "Strange Magic, Sweet Talkin' Woman, Turn to Stone",
    "9641": "Confusion, Last Train to London, All Over the World, Hold On Tight",
    "2853": "Can't Get It Out of My Head, Showdown, Twilight",
    "6704": "Ticket to the Moon, Here Is the News, Twilight, Secret Messages, Four Little Diamonds"
}

In [14]:
users = {
    "4589": "Eve",
    "7312": "Diana",
    "9641": "Alice",
    "2853": "Charlie",
    "6704": "Bob"
}


In [15]:
for user_id, song_list in songs.items():
    song_counts= len(song_list.split(", "))

In [16]:
song_counts

5

In [17]:
# Creating a list to store the names of the people who listened to more than one song
multiple_song_list = []

#Iterate over the songs dict
for user_id, song_list in songs.items():
    #Split the song list into individual songs and counting them
    song_counts= len(song_list.split(", "))
    
    #If statement to check if the user listened to more than one song
    if song_counts > 2:
        user_name = users[user_id]
        
        multiple_song_list.append(user_name)
        
print(f"The people who listened to more than one song are: {multiple_song_list}")

The people who listened to more than one song are: ['Eve', 'Diana', 'Alice', 'Charlie', 'Bob']


**7.1** What is the name of the user who listened to the most songs?

*Tip: You will want to check the methods associated with string manipulation. See the link: <br>
https://python-reference.readthedocs.io/en/latest/docs/unicode/index.html*

In [18]:
#Initializing variables to store the user with most songs and the maximum song count
max_songs= 0
max_songs_user= ""

#Using a for loop over the songs dictionary 
for user_id, song_list in songs.items():
    #As we did on the 7., split the song list into individual songs and counting them
    song_counts= len(song_list.split(", "))
    
    #Write an if statement to get the maximum number of songs and them extract the user
    if song_counts > max_songs:
        max_songs = song_counts
        max_songs_user = users[user_id]
    
print(f"The user who listened to the most songs is {max_songs_user} who listened {max_songs} songs")   
        

The user who listened to the most songs is Bob who listened 5 songs


**7.2** What is the name of the most popular song?

In [19]:
#Creating a dictionary to store the song counts
song_counts= {}

#Using a for loop over the songs dictionary 
for song_list in songs.values():
    #As we did on the 7., split the song list into individual songs
    songs_list_split = song_list.split(", ")
    
    for song in songs_list_split:
        if song in song_counts:
            song_counts[song] += 1
        else:
            song_counts[song] = 1
            
#Finding the song with the max counting
max_song = max(song_counts, key=song_counts.get)

print(f"The most popular song is {max_song}")
    

The most popular song is Twilight


# <span style="color:brown"> Part 2 -  Data loading and analysis  </span>

## Exercise I - Alice what do you have to say?

**8**: Load the Alice text file into a variable called Alice. <br>

*Note: Use a relative filepath in relation to the location of your notebook.*

**9**: Create a list in which each element is a word from the file *Alice*. Store that list in a variable called *wAlice*. <br>

*Note: You will need to do some text parsing here. In particular to split the sentences into words. It is also a good practice to normalize words so that words "Hello" and "hello" become identical, by making all letters lower case. <br>*

*Tip: check the following links for a discussion on regular expressions. Also consult the methods available for string manipulation.*

https://docs.python.org/3/library/re.html

https://stackoverflow.com/questions/1276764/stripping-everything-but-alphanumeric-chars-from-a-string-in-python

##### Using the list *wAlice* answer the following questions:<br>
**10**: How many words are in the file Alice.txt? <br>

**11**: What is the longest and smallest word in the text file?

*Note: Length in this case is measured in terms of the number of characters.*

**12**: Delete all the repeated words from *wAlice*. 

**12.1**:How many different words does the text contain?

## Exercise II - Of Countries I Love

Consider the list countries in the cell below. <br>
It consists of a list of the 3-digit ISO codes of a set of countries of interest.

**Note**: If you find the below list of tuples to be erroneous, correct it accordingly and comment on your correction.

In [20]:
countries = [

('PRT','Portugal','Europe'),
('ISL','Israel','Asia'),
('COL','Colombia','South America'),
('NEV','Nevada','North America'),
('JPN','Japan','Asia'),
('RUS','Russia','well... which part?'), 
('DEN','Denmark','Europe'),
('NOR','Norway','Europe')
]

**Now consider the file *cdata.csv* that you should download.** <br>

The file contains several information about countries, and is organized as follows:
1. Column 1 is the 3-digit **ISO Code**
2. Column 2 is the **Full Name** of the country
3. Column 3 is the **Continent** of the country
4. Column 4 is the **Population Size** in 2010
5. Column 5 is the **GDP per capita** in 2010

**13**: Using the Library CSV from Python STL, load the file *cdata.csv* into an object called *raw*. 

**13.1**: Create a dictionary called *cData* in which the key corresponds to the 3-digit ISO Code and the value is a tuple with the information contained in the 2nd to the 5th of column of the cdata file.

**14**: Using *cData*, identify what is the most common Continent among the nations in the list *countries*.

**15**: Using *cData*, identify what is the the most populated nation in the list *countries*.

**16**: Compare the average GDP per capita of the nations in the list *countries* with the average GDP of the countries in cdata file. What can you conclude?

# <span style="color:brown"> Part 3 - Functions hurt nobody </span>

## Exercise I - I hate math

##### Consider the following equation:

\begin{equation}
    y = 6x^2 + 3x + 2 + cos(x)
\end{equation}

**17**: Write a function called *f* that takes one argument, x, and returns y according to the equation above. *Note: You should use the module math to solve this exercise.*

In [21]:
# Import the math module
import math

In [22]:
# Define the function f
def f(x):
    return 6 * x**2 + 3*x + 2 + math.cos(x)

**17.1**: Call the function for x = 2 and print the answer.

In [23]:
# Call the function for x = 2
answer = f(2)

# Print the answer
print(answer)

31.583853163452858


##### Consider the following sequence of numbers:

\begin{equation}
    x_0 = 0;  \\
    x_{n} = x_{n-1} - n; ~~~ \textit{ if} ~~~ x_{n-1}-n > 0 ~~~ \textit{ and is not already in the sequence}\\
    x_{n} = x_{n-1} + n; ~~~ \textit{ otherwise} ~~~ ~~~ ~~~ ~~~ ~~ ~~~ ~~~ ~~~ ~~~ ~~~ ~ ~~~ ~~~ ~~~ ~~~ ~~~ ~~~ ~ ~~~ ~~~ ~~~~ ~~~ ~~~ ~~~ ~ \\
\end{equation}

**17.2**: Write a function that returns the nth digit of the above defined sequence. <br>

*Note: the above sequence is also known as the Recamán's sequence, and it was invented by Bernardo Recamán Santos (Bogotá, Colombia)*

In [24]:
# Define the function RecamanSequence
def RecamanSequence(n: int) -> int:
    """
    Returns the nth digit in Recamán's sequence.

    Args:
    - n (int): The 1-based index of the digit in the sequence to return. Must be a positive integer.

    Returns:
    - int: The nth digit of the Recamán's sequence.
    """
        
    # Raise an error if n is not positive
    if n <= 0:
        raise ValueError("n must be a positive integer!")
    
    # Initialize the sequence with 0 as the first digit
    sequence = [0]

    # Create the sequence until the nth digit
    for i in range(1, n):
        # Subtract the current index from the last digit
        value = sequence[i-1] - i
        
        # If the value is positive and not already in the sequence, append it
        if value > 0 and value not in sequence:
            sequence.append(value)
        # Otherwise, add the index to the last digit and append it
        else:
            sequence.append(sequence[i-1] + i)
    
    # Return the nth digit in the sequence
    return sequence[n-1]

In [25]:
# Test the RecamanSequence function by printing the first 10 digits of Recamán's sequence
for n in range(1, 11):
    print(f"x{n} = {RecamanSequence(n)}")

x1 = 0
x2 = 1
x3 = 3
x4 = 6
x5 = 2
x6 = 7
x7 = 13
x8 = 20
x9 = 12
x10 = 21


**17.3** Write a function named isPrime that takes an integer as input and outputs True if the number is prime and False if the number if not prime. 

In [26]:
# Define the function isPrime
def isPrime(x: int) -> bool:
    """
    Check if a number is prime.

    Args:
    - x (int): The number to check for primality. Must be a positive integer.

    Returns:
    - bool: True if x is prime, False otherwise.
    """
    
    # Raise an error if x is not positive
    if x <= 0:
        raise ValueError("x must be a positive integer!")
    
    # Handle the case for 1 separately (1 is not a prime number)
    if x == 1:
        return False
    
    # Iterate from 2 to x-1 to check for divisors
    for i in range(2, x):
        # If x is divisible by any number in this range, it's not a prime number
        if x % i == 0:
            return False
    
    # If no divisors were found, x is prime
    return True

**17.4** Test your function against some examples to show that it works as expected.

In [27]:
# Test the isPrime function by printing the primality of the numbers from 1 to 10
for i in range(1, 11):
    print(f"{i}: {isPrime(i)}")

1: False
2: True
3: True
4: False
5: True
6: False
7: True
8: False
9: False
10: False


## Exercise II - Monty Hall Problem


**18**: The Monty Hall Problem is a probability puzzle loosely based on the American television game show "Let's Make a Deal" and named after its original host, Monty Hall. The puzzle can be stated as follows

> Suppose you're on a game show, and you're given the choice of three doors: Behind one door is a car; behind the others, goats. You pick a door, say No. 1, and the host, who knows what's behind the doors, opens another door, say No. 3, which has a goat. He then says to you, "Do you want to pick door No. 2?" Is it to your advantage to switch your choice? <br>
+info: https://en.wikipedia.org/wiki/Monty_Hall_problem

While intuitively your guess might be that the chances of winning the car is 1/3 independently of the choice to switch doors or not. However, the odds of winning if you switch doors are 2/3 and greater than if you decide not to switch. While there are theoretical solutions to this problem that allow us to estimate the correct odds of winning in each scenario -- switching or not doors -- an alternative way to proof such outcome is through computer simulations. That is the goal of this exercise, to write the necessary components to simulate the Monty Hall problem and validate the theoretical results through simulations.

In this exercise, start by implementing a function that simulates one instance of the Monty Hall problem. 
In that sense, write a function called MontyHall that accepts **one argument**:
- A boolean (True/False, or 1/0) that specifies if the player switches doors or not (after the host has opened his door, which contains a goat).
The function should return True if the player wins the car or False if not.

Naturally, for the function to reproduce the contest it needs to consider and perform some actions, such as:
- Set up the game, that is, create the necessary variables to store the information about three doors and what they have behind, as well as an indicator to track the choice of the player.
- Given the initial choice of the player, simulate the opening of one of the two remaining doors that contains a goat.
- Given the last two doors left, simulate if the player wants to switch or not his/her choice given their strategy (defined by the input boolean argument of the function).
- Output the result, if the contestant correctly guesses which door hides the car or if not.

The function MontyHall should simulate one instance of the contest given the choice of the player (to switch or not).

**The goal is to understand, statistically, which action leads to the highest probability of winning.** <br>
*Note: You should use the library random to solve this exercise.*

In [28]:
# Import the random module
import random

In [29]:
# Define the function MontyHall
def MontyHall(switch: bool) -> bool:
    """
    Simulate one instance of the Monty Hall problem.

    Args:
    - switch (bool): True if the player wants to switch doors after the host opens a door, which contains a goat, False otherwise.

    Returns:
    - bool: True if the player correctly guesses which door hides the car, False otherwise.
    """

    # Initialize the doors: one with a car, two with a goat
    doors = ["car", "goat", "goat"]

    # Shuffle the doors randomly
    random.shuffle(doors)

    # Simulate the player randomly selecting one of the three doors
    player_door = random.randint(0, 2)

    # Identify which doors were not chosen by the player
    remaining_doors = [i for i in range(3) if i != player_door]
    
    # Simulate the host opening a door, which contains a goat
    host_door = next(i for i in remaining_doors if doors[i] == "goat")
    
    # Simulate if the player switches doors
    if switch:
        # Remove the door opened by the host
        remaining_doors.remove(host_door)
        
        # Switch to the remaining door
        player_door = remaining_doors[0]

    # Return True if the player's final choice contains the car
    return doors[player_door] == "car"

In [30]:
# Test the MontyHall function with the player choosing to switch doors
MontyHall(True)

True

**18.1** Now that you have a function that simulates an instance of the Monty Hall problem, implement an experiment where you repeat many times (thousands of times) for each of the two possible scenarios: player switching the door, and player not switching the door. Keep track of the results and estimate the frequencies of wins for each scenario and discuss if the results are inline with the theoretical odds.

In [31]:
# Define the function ExperimentMontyHall
def ExperimentMontyHall(n: int):
    """
    Run an experiment to simulate the Monty Hall problem n times for each of the two possible scenarios:
    player switching the door, and player not switching the door.
    Calculate the win percentage for both scenarios and print the results.
    
    Args:
    - n (int): The number of simulations to run for each scenario (switching or not switching).

    Returns:
    - None: The function prints the win percentages for both scenarios, but does not return any value.
    """
    
    # Initialize a list to store the results of the simulations for both scenarios
    scenarios_simulations = []

    # Simulate the two possible scenarios: switching or not switching
    for switch in [True, False]:
        # Initialize a list to store the result of each simulation for the current scenario
        simulations = []

        # Simulate the Monty Hall problem n times for the current scenario, and append the results to "simulations"
        for _ in range(n):
            simulations.append(MontyHall(switch))
        
        # Append the list of results for the current scenario to "scenarios_simulations"
        scenarios_simulations.append(simulations)
    
    # Calculate the win percentage when the player switches doors
    switching_wins_percentage = sum(scenarios_simulations[0]) / n * 100
    
    # Calculate the win percentage when the player doesn't switch doors
    not_switching_wins_percentage = sum(scenarios_simulations[1]) / n * 100

    # Print the results
    print("Wins percentage")
    print(f"Switching scenario: {round(switching_wins_percentage, 2)}%")
    print(f"Not switching scenario: {round(not_switching_wins_percentage, 2)}%")

In [32]:
# Test the ExperimentMontyHall function with 10,000 simulations for each scenario
ExperimentMontyHall(10000)

Wins percentage
Switching scenario: 66.97%
Not switching scenario: 33.37%


For the switching and not switching scenarios, the win percentages in the simulation are 66.97% and 33.37%, respectively. These results are very close to the theoretical odds of 2/3 and 1/3, respectively.<br>

**Conclusion:** The results from 10,000 simulations align well with the theoretical odds of the Monty Hall problem. If the player chooses to switch doors, they significantly increase their chance of winning (around 2/3 of the time). On the other hand, if they stick with their original choice, their chances of winning are much lower (around 1/3 of the time).

## Exercise III

**19**: Consider the list *A*, declared below. <br>
Use the function map() to change the values of A by adding 1 to each value if it is even and subtracting 1 from it otherwise.

**Note**: Carefully consider the values of list A and adapt your code accordingly to deal with number that are represented as strings.

In [33]:
A = [460,3347,3044,490,699,"1258",1804,973,2223,3416,"2879",1058,2915,2422,351,"1543",1020,208,643,"795",3337,"2585",471,2623,1077]

In [34]:
# For each element of A (x), convert it to an integer:
# - If x is even, add 1 to it
# - If x is odd, subtract 1 from it
A = list(map(lambda x: int(x)+1 if int(x)%2 == 0 else int(x)-1, A))

In [35]:
# Print the modified list A
print(A)

[461, 3346, 3045, 491, 698, 1259, 1805, 972, 2222, 3417, 2878, 1059, 2914, 2423, 350, 1542, 1021, 209, 642, 794, 3336, 2584, 470, 2622, 1076]


**20**: Create a list B with the same size of A, where each element is True if the associated value in list A is greater than 700, else is False.

In [36]:
# Create the list B
B = list(map(lambda x: x>700, A))

In [37]:
# Print the list B
print(B)

[False, True, True, False, False, True, True, True, True, True, True, True, True, True, False, True, True, False, False, True, True, True, False, True, True]


**21**: Create a list L that contains the Logarithm of base 10 of each value in A. <br>

*Note: You should use the module math to solve this exercise.*

In [38]:
# Import the math module
import math

In [39]:
# Create the list L
L = list(map(lambda x: math.log10(x), A))

In [40]:
# Print the list L
print(L)

[2.663700925389648, 3.5245259366263757, 3.4835872969688944, 2.6910814921229687, 2.843855422623161, 3.1000257301078626, 3.256477206241677, 2.9876662649262746, 3.346744054604849, 3.5336449787987627, 3.4590907896005865, 3.024895960107485, 3.4644895474339714, 3.384353414137506, 2.5440680443502757, 3.188084373714938, 3.0090257420869104, 2.3201462861110542, 2.807535028068853, 2.8998205024270964, 3.523226041965701, 3.4122925093230463, 2.6720978579357175, 3.4186326873540653, 3.0318122713303706]


**22**: How many numbers in A are greater than 1000? <br>

*Note: You should use the function filter().*

In [41]:
# Create a filtered list with the elements of A that are greater than 1000
numbers_above_1000 = list(filter(lambda x: x>1000, A))

In [42]:
# Print the length of the filtered list (the number of elements of A that are greater than 1000)
print(len(numbers_above_1000))

16


**23** Use the function isPrime, that you declared in 17.3, to identify which numbers from A are prime. Note that depending on your implementation of isPrime this task can take more or less time. Use filter to achieve this task.

In [43]:
# Create a filtered list with the elements of A that are prime
prime_numbers = list(filter(lambda x: isPrime(x), A))

In [44]:
# Print the filtered list (the elements of A that are prime)
print(prime_numbers)

[461, 491, 1259, 2423, 1021]


### Congratulations, it is done!